### PR TITLE
Remove some dead code

### PR DESCRIPTION
**Pull request by asottile**
- I wrote a thing: https://github.com/asottile/dead
- wanted to try it out, there's lots of false positives and I didn't look
  through all the things it pointed out but here's some

### CODE REVIEW

File: src/_pytest/junitxml.py [hunk 1]
```diff
@@ -27,10 +28,6 @@
 # Python 2.X and 3.X compatibility
 if sys.version_info[0] < 3:
     from codecs import open
```
Comment (RonnyPfannschmidt) [inline src/_pytest/junitxml.py]: i believe it should be possible to go for `io.open` by now

File: src/_pytest/fixtures.py [hunk 2]
```diff
@@ -660,12 +660,6 @@ def addfinalizer(self, finalizer):
         self._fixturedef.addfinalizer(finalizer)
 
 
-class ScopeMismatchError(Exception):
```
Comment (RonnyPfannschmidt) [inline src/_pytest/fixtures.py]: this one should be in actual use, what happened there

File: src/_pytest/assertion/rewrite.py [hunk 3]
```diff
@@ -51,19 +51,6 @@ def ast_Call(a, b, c):
         return ast.Call(a, b, c, None, None)
 
 
-def ast_Call_helper(func_name, *args, **kwargs):
```
Comment (RonnyPfannschmidt) [inline src/_pytest/assertion/rewrite.py]: for that one i'd like to know why it was and why it isnt

File: src/_pytest/_code/source.py [hunk 4]
```diff
@@ -77,17 +77,6 @@ def strip(self):
         source.lines[:] = self.lines[start:end]
         return source
 
-    def putaround(self, before="", after="", indent=" " * 4):
```
Comment (RonnyPfannschmidt) [inline src/_pytest/_code/source.py]: those methods are potentially breaking accidential  external api, we should document that as removal

File: testing/code/test_source.py [hunk 5]
```diff
@@ -149,7 +107,6 @@ def g(x):
 
     def test_getrange(self):
         x = self.source[0:2]
-        assert x.isparseable()
```
Comment (RonnyPfannschmidt) [inline testing/code/test_source.py]: ouch

File: src/_pytest/fixtures.py (hunk 2 above)
Comment (asottile) [reply src/_pytest/fixtures.py]: 3ed3e51997b30f2d739515c7daa33f086379e182 was the last usage removed

File: testing/python/metafunc.py [hunk 6]
```diff
@@ -784,7 +784,6 @@ def pytest_generate_tests(metafunc):
 
             @pytest.fixture
             def metafunc(request):
-                assert request._pyfuncitem._genid == "0"
```
Comment (RonnyPfannschmidt) [inline testing/python/metafunc.py]: now that one is interesting

File: src/_pytest/assertion/rewrite.py (hunk 3 above)
Comment (asottile) [reply src/_pytest/assertion/rewrite.py]: last usage removed in 3e6f1fa2db20eb2fd751a778dc42dd126a9e778d

File: src/_pytest/junitxml.py (hunk 1 above)
Comment (asottile) [reply src/_pytest/junitxml.py]: yeah I thought about this, but decided to leave it for another PR given `codecs.open` works more like python2.x open:

```pycon
>>> from codecs import open
>>> with open('f', 'w', encoding='utf-8') as f:
...     f.write(b'wat')
... 
>>> import io
>>> with io.open('f', 'w', encoding='utf-8') as f:
...     f.write(b'wat')
... 
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
TypeError: write() argument 1 must be unicode, not str
```

File: src/_pytest/_code/source.py (hunk 4 above)
Comment (asottile) [reply src/_pytest/_code/source.py]: 👍 definitely, I'll make sure to list these out tomorrow in a changelog entry -- already targeting features

File: testing/python/metafunc.py (hunk 6 above)
Comment (asottile) [reply testing/python/metafunc.py]: it is interesting!  here's the last usage of that one: 426907eafbba7738cf33374a33c3c7aed67dadce

File: src/_pytest/fixtures.py (hunk 2 above)
Comment (RonnyPfannschmidt) [reply src/_pytest/fixtures.py]: yikes, its bad that this was missed, but now its way too late ^^

File: src/_pytest/junitxml.py (hunk 1 above)
Comment (RonnyPfannschmidt) [reply src/_pytest/junitxml.py]: most interesting, but also something we should unify towards the unicode only behaviour in junitxml

File: src/_pytest/nodes.py [hunk 7]
```diff
@@ -217,9 +217,6 @@ def listextrakeywords(self):
             extra_keywords.update(item.extra_keyword_matches)
         return extra_keywords
 
-    def listnames(self):
```
Comment (RonnyPfannschmidt) [inline src/_pytest/nodes.py]: external api breach of node

File: testing/python/metafunc.py (hunk 6 above)
Comment (RonnyPfannschmidt) [reply testing/python/metafunc.py]: yikes, thanks

File: src/_pytest/_code/source.py (hunk 4 above)
Comment (Tadaboody) [reply src/_pytest/_code/source.py]: Maybe give a depreciation grace period so we don't have as many breaks as the recent mark changes?

Comment (nicoddemus) [reply src/_pytest/_code/source.py]: I agree, IMO we should avoid further breakages. 👍 for adding a deprecation warning.

File: src/_pytest/_code/source.py [hunk 8]
```diff
@@ -118,26 +107,6 @@ def deindent(self):
         newsource.lines[:] = deindent(self.lines)
         return newsource
 
-    def isparseable(self, deindent=True):
```
Comment (nicoddemus) [inline src/_pytest/_code/source.py]: Ditto about deprecating it first.

File: src/_pytest/junitxml.py [hunk 9]
```diff
@@ -27,10 +28,6 @@
...
-else:
```
Comment (nicoddemus) [inline src/_pytest/junitxml.py]: Great!

**Comment by asottile** [comment]
been a bit busy, I put back the potentially public things and I can follow up with those later, don't really have time to document all the deprecations right now but wanted to get at least this bit out

**Comment by nicoddemus** [comment]
https://github.com/pytest-dev/pytest/pull/4643 fixes the deprecation error.

**Pull request merged**